src/functionsDeployHelper.ts

```diff
@@ -161,10 +184,124 @@ export function getFunctionsInfo(parsedTriggers: CloudFunctionTrigger[], project
         "functions",
         trigger.name,
       ].join("/");
-      functionsInfo.push(triggerDeepCopy);
+      if (!_.get(regionMap, region)) {
+        regionMap[region] = [];
+      }
+      regionMap[region].push(triggerDeepCopy);
     });
   });
-  return functionsInfo;
+  return regionMap;
+}
+
+/**
+ * Helper method to turn a RegionMap into a flat list of all functions in a deployment.
+ * @param regionMap A RegionMap for the deployment.
+ */
+export function flattenRegionMap(regionMap: RegionMap): CloudFunctionTrigger[] {
+  return _.chain(regionMap)
+    .map((value: CloudFunctionTrigger[]) => {
+      return value;
+    })
+    .flatten()
+    .value();
+}
+
+export interface RegionalDeployment {
+  region: string;
+  sourceToken?: string;
+  firstFunctionDeployment?: CloudFunctionTrigger;
+  functionsToCreate: CloudFunctionTrigger[];
+  functionsToUpdate: CloudFunctionTrigger[];
+  schedulesToCreateOrUpdate: CloudFunctionTrigger[];
+}
+
+export interface DeploymentPlan {
+  regionalDeployments: RegionalDeployment[];
+  functionsToDelete: string[];
+  schedulesToDelete: string[];
+}
+
+/**
+ * Create a plan for deploying all functions in one region.
+ * @param region The region of this deployment
+ * @param functionsInLocalSource The functions present in the code currently being deployed.
+ * @param existingFunctionNames The names of all functions that already exist.
+ * @param existingScheduledFunctionNames The names of all schedules functions that already exist.
+ * @param filters The filters, passed in by the user via  `--only functions:`
+ */
+export function createDeploymentPlan(
+  functionsInLocalSource: RegionMap,
+  existingFunctions: CloudFunctionTrigger[],
+  filters: string[][]
+): DeploymentPlan {
+  const deployment: DeploymentPlan = {
+    regionalDeployments: [],
+    functionsToDelete: [],
+    schedulesToDelete: [],
+  };
+  for (const region of Object.keys(functionsInLocalSource)) {
```

Comment:
```suggestion
  for (const region in functionsInLocalSource) {
```

`for...in` is equivalent to `for...of Object.keys`

Reply:
Originally used for.. of here because of the following eslint error: ``` 112:3  error  The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype  guard-for-in``` - we maybe want to get rid of that rule?

Ended up switching to for... in and adding a comment to disable the rule for this line.